File: frontend-next/src/lib/seo/structured-data.ts
/**
 * Brand Battle — Schema.org Structured Data Generator
 * Builds validated JSON-LD graphs for Product, Offer, AggregateOffer, FAQPage,
 * BreadcrumbList, Organization, and WebSite schemas.
 *
 * CRITICAL TRUST RULE: NEVER generate fake ratings, fake reviews, or fabricated attributes.
 * If data is missing, omit property.
 */

import { SEO_CONFIG } from './seo-config'
import { BreadcrumbItem } from './breadcrumbs'

export interface ProductJsonLdProps {
  id: number
  name: string
  description?: string
  image?: string
  brand?: string
  category?: string
  price?: number
  originalPrice?: number
  bestPlatform?: string
  priceVerifiedAt?: string | Date
  currency?: string
  rating?: number
  totalReviews?: number
  url: string
}

export interface ComparisonJsonLdProps {
  product1: ProductJsonLdProps
  product2: ProductJsonLdProps
  canonicalUrl: string
  breadcrumbs: BreadcrumbItem[]
  faqs?: Array<{ question: string; answer: string }>
}

/**
 * Builds Schema.org Product payload.
 */
export function buildProductJsonLd(props: ProductJsonLdProps): Record<string, any> {
  const schema: Record<string, any> = {
    '@context': 'https://schema.org',
    '@type': 'Product',
    name: props.name,
    url: props.url,
  }

  if (props.description) {
    schema.description = props.description
  }

  if (props.image) {
    schema.image = props.image
  }

  if (props.brand) {
    schema.brand = {
      '@type': 'Brand',
      name: props.brand,
    }
  }

  if (props.category) {
    schema.category = props.category
  }

  // Real verified price offer (only if price > 0)
  if (props.price && props.price > 0) {
    schema.offers = {
      '@type': 'Offer',
      price: props.price,
      priceCurrency: props.currency || 'INR',
      availability: 'https://schema.org/InStock',
      url: props.url,
      seller: {
        '@type': 'Organization',
        name: props.bestPlatform || 'Verified Retailer',
      },
    }
    if (props.priceVerifiedAt) {
      schema.offers.priceValidUntil = new Date(
        new Date(props.priceVerifiedAt).getTime() + 86400 * 7 * 1000
      ).toISOString().split('T')[0]
    }
  }

  // Real aggregate rating ONLY if real reviews exist
  if (props.rating && props.rating > 0 && props.totalReviews && props.totalReviews > 0) {
    schema.aggregateRating = {
      '@type': 'AggregateRating',
      ratingValue: props.rating,
      reviewCount: props.totalReviews,
      bestRating: 5,
      worstRating: 1,
    }
  }

  return schema
}

/**
 * Builds Schema.org BreadcrumbList payload.
 */
export function buildBreadcrumbJsonLd(items: BreadcrumbItem[]): Record<string, any> {
  return {
    '@context': 'https://schema.org',
    '@type': 'BreadcrumbList',
    itemListElement: items.map((item, index) => {
      const element: Record<string, any> = {
        '@type': 'ListItem',
        position: index + 1,
        name: item.name,
      }
      if (item.url) {
        element.item = item.url
      }
      return element
    }),
  }
}

/**
 * Builds Schema.org Organization payload.
 */
export function buildOrganizationJsonLd(): Record<string, any> {
  return {
    '@context': 'https://schema.org',
    '@type': 'Organization',
    name: SEO_CONFIG.organizationName,
    url: SEO_CONFIG.domain,
    logo: SEO_CONFIG.organizationLogo,
    contactPoint: {
      '@type': 'ContactPoint',
      email: SEO_CONFIG.supportEmail,
      contactType: 'customer support',
    },
  }
}

/**
 * Builds Schema.org WebSite payload with SearchAction.
 */
export function buildWebSiteJsonLd(): Record<string, any> {
  return {
    '@context': 'https://schema.org',
    '@type': 'WebSite',
    name: SEO_CONFIG.siteName,
    url: SEO_CONFIG.domain,
    potentialAction: {
      '@type': 'SearchAction',
      target: `${SEO_CONFIG.domain}/search?q={search_term_string}`,
      'query-input': 'required name=search_term_string',
    },
  }
}

/**
 * Builds comparison page Schema.org JSON-LD graph.
 */
export function buildComparisonGraphJsonLd({
  product1,
  product2,
  canonicalUrl,
  breadcrumbs,
  faqs,
}: ComparisonJsonLdProps): Record<string, any> {
  const graphNodes: any[] = [
    {
      '@type': 'WebPage',
      '@id': canonicalUrl,
      url: canonicalUrl,
      name: `${product1.name} vs ${product2.name} Comparison`,
    },
    buildBreadcrumbJsonLd(breadcrumbs),
    buildProductJsonLd(product1),
    buildProductJsonLd(product2),
  ]

  if (faqs && faqs.length > 0) {
    graphNodes.push({
      '@context': 'https://schema.org',
      '@type': 'FAQPage',
      mainEntity: faqs.map(faq => ({
        '@type': 'Question',
        name: faq.question,
        acceptedAnswer: {
          '@type': 'Answer',
          text: faq.answer,
        },
      })),
    })
  }

  return {
    '@context': 'https://schema.org',
    '@graph': graphNodes,
  }
}

File: backend/seo_platform/structured_data.py
"""
Brand Battle — 6. Structured Data (Schema.org JSON-LD) Engine & 7. Breadcrumb Navigation
Generates comprehensive Schema.org JSON-LD schemas (Product, Offer, FAQPage, BreadcrumbList, WebPage).
"""

from typing import Dict, Any, List
from models import Product
from seo_platform.config import seo_config
from seo_platform.schemas import FAQItemSchema


class StructuredDataEngine:
    """Generates Schema.org compliant JSON-LD structured data."""

    def generate_json_ld(
        self, p1: Product, p2: Product, canonical_slug: str, faqs: List[FAQItemSchema]
    ) -> Dict[str, Any]:
        """Generate combined Schema.org JSON-LD graph."""
        name1 = f"{p1.brand.name if p1.brand else ''} {p1.name}".strip()
        name2 = f"{p2.brand.name if p2.brand else ''} {p2.name}".strip()
        canonical_url = f"{seo_config.domain}/compare/{canonical_slug}"
        cat_name = p1.category.name if p1.category else "Products"

        faq_elements = [
            {
                "@type": "Question",
                "name": item.question,
                "acceptedAnswer": {
                    "@type": "Answer",
                    "text": item.answer,
                },
            }
            for item in faqs
        ]

        graph = [
            {
                "@context": "https://schema.org",
                "@type": "WebPage",
                "@id": canonical_url,
                "url": canonical_url,
                "name": f"{name1} vs {name2} Comparison",
                "isPartOf": {
                    "@type": "WebSite",
                    "name": seo_config.site_name,
                    "url": seo_config.domain,
                },
            },
            {
                "@context": "https://schema.org",
                "@type": "BreadcrumbList",
                "itemListElement": [
                    {"@type": "ListItem", "position": 1, "name": "Home", "item": seo_config.domain},
                    {"@type": "ListItem", "position": 2, "name": "Compare", "item": f"{seo_config.domain}/compare"},
                    {"@type": "ListItem", "position": 3, "name": cat_name, "item": f"{seo_config.domain}/discover?category={cat_name}"},
                    {"@type": "ListItem", "position": 4, "name": f"{name1} vs {name2}", "item": canonical_url},
                ],
            },
            {
                "@context": "https://schema.org",
                "@type": "FAQPage",
                "mainEntity": faq_elements,
            },
            {
                "@context": "https://schema.org",
                "@type": "Product",
                "name": name1,
                "offers": {
                    "@type": "Offer",
                    "price": getattr(p1, "current_best_price", 0.0) or 0.0,
                    "priceCurrency": "INR",
                    "availability": "https://schema.org/InStock",
                },
            },
            {
                "@context": "https://schema.org",
                "@type": "Product",
                "name": name2,
                "offers": {
                    "@type": "Offer",
                    "price": getattr(p2, "current_best_price", 0.0) or 0.0,
                    "priceCurrency": "INR",
                    "availability": "https://schema.org/InStock",
                },
            },
        ]

        return {"@context": "https://schema.org", "@graph": graph}

    def generate_product_json_ld(self, p: Product) -> Dict[str, Any]:
        """Generate single Schema.org Product JSON-LD without fabricated reviews."""
        name = f"{p.brand.name if p.brand else ''} {p.name}".strip()
        price = getattr(p, "current_best_price", 0.0) or 0.0
        
        product_schema: Dict[str, Any] = {
            "@context": "https://schema.org",
            "@type": "Product",
            "name": name,
            "url": f"{seo_config.domain}/product/{p.id}",
        }

        if getattr(p, "image_url", None):
            product_schema["image"] = [p.image_url]

        if getattr(p, "description", None):
            product_schema["description"] = p.description

        if p.brand:
            product_schema["brand"] = {
                "@type": "Brand",
                "name": p.brand.name,
            }

        if price > 0:
            product_schema["offers"] = {
                "@type": "Offer",
                "price": price,
                "priceCurrency": "INR",
                "availability": "https://schema.org/InStock",
                "url": f"{seo_config.domain}/product/{p.id}",
            }

        # Zero fake reviews rule: only include aggregateRating if genuine reviews exist
        if getattr(p, "average_rating", None) and getattr(p, "total_reviews", None) and p.total_reviews > 0:
            product_schema["aggregateRating"] = {
                "@type": "AggregateRating",
                "ratingValue": p.average_rating,
                "reviewCount": p.total_reviews,
                "bestRating": 5,
                "worstRating": 1,
            }

        return product_schema

    def generate_breadcrumbs(self, p1: Product, p2: Product) -> List[Dict[str, str]]:
        name1 = f"{p1.brand.name if p1.brand else ''} {p1.name}".strip()
        name2 = f"{p2.brand.name if p2.brand else ''} {p2.name}".strip()
        cat_name = p1.category.name if p1.category else "Products"

        return [
            {"name": "Home", "url": "/"},
            {"name": "Compare", "url": "/compare"},
            {"name": cat_name, "url": f"/discover?category={cat_name}"},
            {"name": f"{name1} vs {name2}", "url": ""},
        ]


# Singleton
structured_data_engine = StructuredDataEngine()
